pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,821 @@
+		if err != nil {
+			return nil, err
+		}
+		return &controlplane.BundleServerAuthConfiguration{
+			APIKey: value,
+		}, nil
+	case crdv1alpha1.BearerToken:
+		value, err := parseAuthValue(secret.Data, secretKeyWithBearerToken)
+		if err != nil {
+			return nil, err
+		}
+		return &controlplane.BundleServerAuthConfiguration{
+			BearerToken: value,
+		}, nil
+	case crdv1alpha1.BasicAuthentication:
+		username, err := parseAuthValue(secret.Data, secretKeyWithUsername)
+		if err != nil {
+			return nil, err
+		}
+		password, err := parseAuthValue(secret.Data, secretKeyWithPassword)
+		if err != nil {
+			return nil, err
+		}
+		return &controlplane.BundleServerAuthConfiguration{
+			BasicAuthentication: &controlplane.BasicAuthentication{
+				Username: username,
+				Password: password,
+			},
+		}, nil
+	}
+	return nil, fmt.Errorf("unsupported authentication type %s", authentication.AuthType)
+}
+
+func (c *Controller) uploadToFileServer(up uploader, uid string, parsedURL *url.URL, serverAuth *controlplane.BundleServerAuthConfiguration, packetFile io.Reader) error {
+	joinedPath := path.Join(parsedURL.Path, c.nodeConfig.Name+"_"+uid+".pcapng")
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// #nosec G106: skip host key check here and users can specify their own checks if needed
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	}
+	return up.upload(parsedURL.Host, joinedPath, cfg, packetFile)
+}
+
+func (c *Controller) uploadPackets(ps *crdv1alpha1.PacketSampling, outputFile afero.File) error {
+	klog.V(2).InfoS("Uploading captured packets for PacketSampling", "name", ps.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload support bundle while getting uploader: %v", err)
+	}
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("failed to upload sampled packets to file server while setting offset: %v", err)
+	}
+	// fileServer.URL should be like: 10.92.23.154:22/path or sftp://10.92.23.154:22/path
+	parsedURL, err := parseUploadUrl(ps.Spec.FileServer.URL)
+	if err != nil {
+		return fmt.Errorf("failed to upload packets while parsing upload URL: %v", err)
+	}
+	triesLeft := uploadToFileServerTries
+	var uploadErr error
+	authentication, err := c.parseBundleAuth(ps.Spec.Authentication)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get authentication defined in the PacketSampling CR", "name", ps.Name, "authentication", ps.Spec.Authentication)
+		return err
+	}
+	for triesLeft > 0 {
+		if uploadErr = c.uploadToFileServer(uploader, string(ps.UID), parsedURL, authentication, outputFile); uploadErr == nil {
+			return nil
+		}
+		triesLeft--
+		if triesLeft == 0 {
+			return fmt.Errorf("failed to upload support bundle after %d attempts", uploadToFileServerTries)
```

Comment:
Looks like this is also copy-paste codes from support bundle. Please check how to share them.

Reply:
updated.